syntax = "proto3";

package gladys.memory;

// Memory Storage Service - Python backend
// Handles persistent storage, embeddings, and complex queries
service MemoryStorage {
    // Store a new episodic event
    rpc StoreEvent(StoreEventRequest) returns (StoreEventResponse);

    // Query events by time range
    rpc QueryByTime(QueryByTimeRequest) returns (QueryEventsResponse);

    // Query events by embedding similarity
    rpc QueryBySimilarity(QueryBySimilarityRequest) returns (QueryEventsResponse);

    // Generate embedding for text
    rpc GenerateEmbedding(GenerateEmbeddingRequest) returns (GenerateEmbeddingResponse);

    // Store/update a heuristic
    rpc StoreHeuristic(StoreHeuristicRequest) returns (StoreHeuristicResponse);

    // Query heuristics by condition match
    rpc QueryHeuristics(QueryHeuristicsRequest) returns (QueryHeuristicsResponse);
}

// Salience Gateway Service - the "amygdala"
// Co-located with Memory per ADR-0001 ยง5.1
// Provides fast salience evaluation for incoming events
service SalienceGateway {
    // Evaluate salience of an event
    // Uses heuristics from Memory + novelty detection from Rust fast path
    rpc EvaluateSalience(EvaluateSalienceRequest) returns (EvaluateSalienceResponse);
}

// --- Events ---

message EpisodicEvent {
    string id = 1;                      // UUID
    int64 timestamp_ms = 2;             // Unix timestamp milliseconds
    string source = 3;                  // Sensor ID
    string raw_text = 4;                // Natural language description
    bytes embedding = 5;                // 384-dim float32 vector (serialized)
    SalienceVector salience = 6;
    string structured_json = 7;         // Domain-specific data as JSON string
    repeated string entity_ids = 8;     // Referenced entity UUIDs
}

message SalienceVector {
    float threat = 1;
    float opportunity = 2;
    float humor = 3;
    float novelty = 4;
    float goal_relevance = 5;
    float social = 6;
    float emotional = 7;                // -1.0 to 1.0
    float actionability = 8;
    float habituation = 9;
}

message StoreEventRequest {
    EpisodicEvent event = 1;
}

message StoreEventResponse {
    bool success = 1;
    string error = 2;
}

message QueryByTimeRequest {
    int64 start_ms = 1;
    int64 end_ms = 2;
    string source_filter = 3;           // Optional: filter by source
    int32 limit = 4;
}

message QueryBySimilarityRequest {
    bytes query_embedding = 1;          // 384-dim float32 vector
    float similarity_threshold = 2;     // Minimum cosine similarity
    int64 time_filter_hours = 3;        // Optional: limit to recent events
    int32 limit = 4;
}

message QueryEventsResponse {
    repeated EpisodicEvent events = 1;
    string error = 2;
}

// --- Embeddings ---

message GenerateEmbeddingRequest {
    string text = 1;
}

message GenerateEmbeddingResponse {
    bytes embedding = 1;                // 384-dim float32 vector
    string error = 2;
}

// --- Heuristics (CBR Design per ยง22) ---

message Heuristic {
    string id = 1;                          // UUID
    string name = 2;                        // Human-readable name

    // Condition: fuzzy matching via embedding
    string condition_text = 3;              // Human-readable, used for embedding generation
    bytes condition_embedding = 4;          // 384-dim float32 vector (serialized)
    float similarity_threshold = 5;         // Min cosine similarity to match (default 0.7)

    // Effects (salience modifiers + actions)
    string effects_json = 6;                // JSONB equivalent - action to take

    // Learning
    float confidence = 7;                   // 0.0-1.0, default 0.5
    float learning_rate = 8;                // TD learning rate, default 0.1

    // Provenance
    string origin = 9;                      // 'built_in', 'pack', 'learned', 'user'
    string origin_id = 10;                  // Pack ID, reasoning trace ID, etc.

    // Tree structure (reserved for future, not implemented in PoC)
    repeated string next_heuristic_ids = 11;
    bool is_terminal = 12;                  // Default true

    // Stats
    int64 last_fired_ms = 13;
    int32 fire_count = 14;
    int32 success_count = 15;

    // Timestamps
    int64 created_at_ms = 16;
    int64 updated_at_ms = 17;
}

message StoreHeuristicRequest {
    Heuristic heuristic = 1;
    bool generate_embedding = 2;            // If true, server generates embedding from condition_text
}

message StoreHeuristicResponse {
    bool success = 1;
    string error = 2;
    string heuristic_id = 3;                // ID of created/updated heuristic
}

message QueryHeuristicsRequest {
    // For embedding-based matching (CBR)
    string query_text = 1;                  // Text to embed and match against conditions
    bytes query_embedding = 2;              // Pre-computed embedding (if available)
    float min_similarity = 3;               // Minimum similarity threshold
    float min_confidence = 4;               // Minimum confidence filter
    int32 limit = 5;                        // Max results to return
}

message QueryHeuristicsResponse {
    repeated HeuristicMatch matches = 1;
    string error = 2;
}

message HeuristicMatch {
    Heuristic heuristic = 1;
    float similarity = 2;                   // Cosine similarity to query
    float score = 3;                        // similarity ร confidence
}

// --- Salience Evaluation ---

message EvaluateSalienceRequest {
    string event_id = 1;                // Event identifier
    string source = 2;                  // Sensor source ID
    string raw_text = 3;                // Event text for novelty/semantic eval
    string structured_json = 4;         // Domain-specific context as JSON
    repeated string entity_ids = 5;     // Referenced entities
    bool skip_novelty_detection = 6;    // Skip embedding-based novelty (for benchmarking)
}

message EvaluateSalienceResponse {
    SalienceVector salience = 1;        // Evaluated salience dimensions
    bool from_cache = 2;                // True if result was cached/heuristic
    string matched_heuristic_id = 3;    // ID of matched heuristic (if any)
    string error = 4;
    bool novelty_detection_skipped = 5; // True if skip_novelty_detection was set
}
